### ISSUE TITLE
Can't install package from repo that is added by mgmt

### ISSUE THREAD

**Issue by wdullaer** [purpleidea/mgmt#503]
## Versions:

* mgmt version (eg: `mgmt --version`): mgmt 0.018

* operating system/distribution (eg: `uname -a`): Ubuntu 19.04

## Description:
You cannot add an apt repository and install a package from it in the same mgmt run.

mgmt will try to list the file contents of the package before it starts executing the resource graph, which will cause the packagekit resolve to fail since at this point the repo hasn't been added yet. This happens for two reasons:

* Auto-edges: https://github.com/purpleidea/mgmt/blob/d5b7dc0acc3b608348c891a3f274f6ee3bad302d/engine/resources/pkg.go#L566 calls populateFileList. This is understandable though, you need to know the files the package provides for AutoEdges to work. Rather than returning an error it might be better to just disable AutoEdges if the package can't be resolved.
* Init: https://github.com/purpleidea/mgmt/blob/d5b7dc0acc3b608348c891a3f274f6ee3bad302d/engine/resources/pkg.go#L90
Init is called before the graph is executed. I don't really understand why this is done at this point rather than at a later point in the lifecycle of the resource.

**Comment by purpleidea** [comment]
If adding an apt repository is not producing an event that we (via packagekit) detect, then that sounds like a bug we need to fix too!

**Comment by purpleidea** [comment]
>  you need to know the files the package provides for AutoEdges to work. Rather than returning an error it might be better to just disable AutoEdges if the package can't be resolved.

If it's definitely caused by the scenario you describe,  then that's what we'll have to do.

> Init is called before the graph is executed

If it turns out we should call it later, then that's fine to change. If we can catch more issues at Validate or Init than at runtime during CheckApply or Watch, then it's good to have, but if it's causing issues because of repo changes, then we'd have to move it of course.

**Comment by wdullaer** [comment]
> If adding an apt repository is not producing an event that we (via packagekit) detect, then that sounds like a bug we need to fix too!

I can't say, since mgmt errors out before it starts executing the graph. For good measure, here's the mcl file I'm trying to run

```
pkg "software-properties-common" {
	state => "newest",
}

exec "docker-key" {
	cmd => "wget -qO- https://download.docker.com/linux/ubuntu/gpg | apt-key add -",
	shell => "/bin/bash",
	ifcmd => "[[ $(apt-key fingerprint 0EBFCD88 | wc -l) -eq 0 ]]",
	ifshell => "/bin/bash",
}

exec "docker-ppa" {
	cmd => "add-apt-repository -y \"deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable\"",
	shell => "/bin/bash",
	ifcmd => "[[ $(apt-cache policy | grep download.docker.com | wc -l) -eq 0 ]]",
        ifshell => "/bin/bash",

	Depend => Pkg["software-properties-common"],
	Depend => Exec["docker-key"],
}

pkg "docker-ce" {
	state => "newest",
	allowunsupported => true, # Anything not in the main repo is considered unsupported by packagekit
	Meta:autoedges => false, # It'll error out even earlier trying to list the files in the packages for AutoEdges
	Depend => Exec["docker-ppa"],
}
```

I'd take a stab at fixing these, but I haven't really had the time so far.

**Comment by purpleidea** [comment]
> 	allowunsupported => true, # Anything not in the main repo is considered unsupported by packagekit

Is that the trick to getting it to install from universe?? IOW that's the fix for https://github.com/purpleidea/mgmt/issues/462 ?

> I'd take a stab at fixing these, but I haven't really had the time so far.

Great! Git master has an important bug fix for import/class scoping. I'm currently working on an important performance fix for type unification that makes it currently unusable with certain code bases. After that I'll circle back and help fix or mentor any misc bugs that are remaining.

**Comment by purpleidea** [comment]
A user on IRC asked about this issue. I haven't updated it with news lately, sorry.

TL;DR: The failures at Init can happen because we pre-populate for AutoEdges... If you don't want AutoEdges we can avoid doing that in Init(). If someone is on hand to test that patch, please ping here.

**Comment by purpleidea** [comment]
I think all of these issues should be resolved. If anyone notices anything please lmk.
